[DOCS-1453] Docs for run status change automations

mdlinville opened the pull request:
[DOCS-1453] Docs for run status change automations

Filed https://wandb.atlassian.net/browse/DOCS-1550 to get an updated graphic as a fast follow.

Preview: https://docs-1453.docodile.pages.dev/guides/core/automations/automation-events/#run-events

Ready for peer review.

[DOCS-1453]: https://wandb.atlassian.net/browse/DOCS-1453?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mdlinville committed:
[DOCS-1453] Docs for run status change automations

ssisk commented on content/guides/core/automations/automation-events.md:
```diff
@@ -60,9 +60,12 @@ From the project's **Automations** tab or directly from a line plot panel, you c
 - A [system metric]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}) such as `cpu`, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds.
 
 The notification can watch the metric for these events:
+- **Run status change**: Trigger a workflow when a run changes its status to one or more of **Running**, **Failed**, or **Finished**.
```
One tricky thing is that these run status change events aren't available when creating the automation in the workspace - it's only available from the automations tab (since it's not related to the specific metric chosen, it's related to the whole run)

ssisk commented on content/guides/core/automations/automation-events.md:
```diff
@@ -60,9 +60,12 @@ From the project's **Automations** tab or directly from a line plot panel, you c
 - **Run metrics change threshold met**: Trigger a workflow when the average logged values of a run change by the absolute or relative threshold you specify.
 
+Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
+
```
It may be worth clarifying what "finished" means in the docs, since the term is ambiguous. it could mean "one way or another it finished" - that could include success or failure. Or it could mean "the run succeeded"

Finished for run status change events means the run succeeded. 

I would have preferred different terminology to avoid this ambiguity in the first place, we went this direction since it matches the terminology in other places in W&B.

mdlinville replied:
I added a bit more detail, but where I talk about the states, I do say that it can be Running, Finished, or Failed. That reads to me like it can't be both finished and failed.

mdlinville committed:
Stephen's feedback
Noah's feedback
Typos

Co-authored-by: Noah Luna <15202580+ngrayluna@users.noreply.github.com>

ssisk commented on content/guides/core/automations/create-automations/slack.md:
```diff
@@ -61,11 +61,13 @@ A Registry admin can create automations in that registry.
+    From a line plot in a workspace, you can quickly create an automation for a [run metric]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
     {{< img src="/images/automations/run_metric_automation_from_panel.png" alt="Illustration showing the location of the automation bell icon" >}}
+
+    You can create a run status automation only from the **Automations** page, not from a workspace. Run status is a property of the entire run, not a single metric. To learn more, refer to [Run events]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}).
```
❤️

ssisk replied:
okay, now that I'm re-reading this, I'm realizing that I really like this paragraph, but it is very odd on the slack and webhook pages - it really belongs on the event page. Also, looking at the events page, I'm not sure what more they would learn from reading the run events page :) 

This whole paragraph is just about run status automations, and doesn't have anything to do with slack/webhook actions.

ssisk commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -64,10 +64,12 @@ A Registry admin can create automations in that registry. Registry automations a
+1. In the sidebar, click **Automations**, then click **Create automation**. From here you can create an automation for a run metric or a run status change.
 
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+    From a line plot in a workspace, you can quickly create an automation for a [run metric]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
```
something I missed when previously reviewing: to create a webhook automation in this scenario, you'll need to switch to the "advanced" configuration mode.

mdlinville replied:
Done, but I find this confusing. If I fill in the metric details before clicking Advanced Configuration, it populates the values in the Advanced form, but that form is set up radically differently from the Slack form. Should I say something like "To configure a webhook automation, or for advanced configuration of a Slack automation, click Advanced Configuration"? I'm not seeing what is Webhook-specific about the Advanced form.

ssisk replied:
You're correct that there is nothing webhook specific, but it is necessary to switch to the advanced view to configure webhooks (since basic doesn't allow you to pick a webhook), and since this is the docs page for webhook automations, if we're going to mention the workspace entry point, I think we should discuss how to use webhooks from that entry point

mdlinville replied:
Got it, thanks for explaining.

mdlinville commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -64,10 +64,12 @@ A Registry admin can create automations in that registry. Registry automations a
+1. In the sidebar, click **Automations**, then click **Create automation**. From here you can create an automation for a run metric or a run status change.
 
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+    From a line plot in a workspace, you can quickly create an automation for a [run metric]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
```
```suggestion
    From a line plot in a workspace, you can quickly create an automation for a [run metric]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel. From here, you can configure a Slack automation by selecting the destination channel. To configure a webhook automation, click **Advanced configuration**.
```

ssisk commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -64,10 +64,12 @@ A Registry admin can create automations in that registry. Registry automations a
+    From a line plot in a workspace, you can quickly create an automation for a [run metric]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel. From here, you can configure a Slack automation by selecting the destination channel. To configure a webhook automation, click **Advan ...
     {{< img src="/images/automations/run_metric_automation_from_panel.png" alt="Illustration showing the location of the automation bell icon" >}}
+
+    You can create a run status automation only from the **Automations** page, not from a workspace. Run status is a property of the entire run, not a single metric. To learn more, refer to [Run events]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}).
```
same comment I made above on the slack page applies here

mdlinville commented on content/guides/core/automations/create-automations/slack.md:
```diff
@@ -61,11 +61,14 @@ A Registry admin can create automations in that registry.
-1. Go the project page and click the **Automations** tab, then click **Create automation**.
-
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+1. Log in to W&B and go to the project page.
```
PTAL

mdlinville commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -64,10 +64,16 @@ A Registry admin can create automations in that registry. Registry automations a
 A W&B admin can create automations in a project.
 
 1. Log in to W&B and go to the project page.
+<<<<<<< HEAD
```
```suggestion
```

mdlinville commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -65,9 +65,14 @@ A W&B admin can create automations in a project.
 1. In the sidebar, click **Automations**, then click **Create automation**.
-
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+    {{% alert %}}
```
PTAL

ssisk commented on content/guides/core/automations/create-automations/webhook.md:
```diff
@@ -65,9 +65,14 @@ A W&B admin can create automations in a project.
     {{< img src="/images/automations/run_metric_automation_from_panel.png" alt="Illustration showing the location of the automation bell icon" >}}
+    Because run status is a property of the entire run, you can create a run status automation from the **Automations** page, not from a workspace.
+
+    To learn more, refer to [Run events]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}).
```
I don't really see how this addresses the concerns I raised:
1. This is my main concern: for someone who is trying to create a webhook automation, it's weird to have this comment here in the middle of that. This information *is* very useful to have on the run events page, so I think putting it here on the actions page makes it harder to find.
2. We don't actually explain more about run status events on the run events page (although the framing now does make it more general than just run status events, but you have a link to the events page just below on line 76, so I'm not sure why we need a link here)

mdlinville replied:
Got it, let me see what I can do. I think I was trying to show them how to quickly create a run metric automation from the line plot, while saying that you can't create run status automations from there. We have separate pages for creating a Slack automation and a webhook automation because PM wanted that, so that means that I need to explain some of these details in multiple places.

mdlinville replied:
OK, let me know if this is nearer the mark.

ssisk reviewed (approved):
looks great!

ngrayluna commented on content/guides/core/automations/automation-events.md:
```diff
@@ -55,48 +55,56 @@ This section describes the events related to an artifact that can trigger an aut
-From the project's **Automations** tab or directly from a line plot panel, you can create a run metric automation triggered by:
-- A metric in a run's history.
-- A [system metric]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}) such as `cpu`, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds.
+An automation be trigered by the following events related to runs.
```
```suggestion
An automation can be triggered by a change in a [run's status]({{< relref "/guides/runs/#run-states" >}}) or a [metric]({{< relref "/guides/track/log/#what-data-is-logged-with-specific-wb-api-calls" }}).
```

ngrayluna commented on content/guides/core/automations/automation-events.md:
```diff
@@ -55,48 +55,56 @@ This section describes the events related to an artifact that can trigger an aut
-- **Run metrics threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged values meets the threshold you specify.
-- **Run metrics change threshold met**: Trigger a workflow when the average logged values of a run change by the absolute or relative threshold you specify.
+#### Run status change
+Trigger a workflow when a run changes its status to one or more of **Running**, **Finished**, or **Failed**. **Finished** indicates a run that completed successfully. Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
```
This makes it sound like a run can have multiple statuses at once. Suggest to something like:

```suggestion
Trigger a workflow when a [run's status]({{< relref "/guides/runs/#run-states" >}}) changes. Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
```

Side note: The table I linked to has "Killed" unsure if that is an option still.

mdlinville replied:
The UI doesn't present "Killed" as an option. @ssisk Is this intentional?

ssisk replied:
Correct, there are many different ways for a run to end, but for purposes of this automation, we grouped them together into failed or finished.

ngrayluna commented on content/guides/core/automations/automation-events.md:
```diff
@@ -55,48 +55,56 @@ This section describes the events related to an artifact that can trigger an aut
+Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
+
+{{% alert %}}
+From a line plot in a workspace, you can quickly create an automation from a line plot in the workspace that shows the metric. Hover over the panel, then click the bell icon at the top of the panel. 
```
```suggestion
From a line plot in a workspace, you can create an automation from a line plot in the workspace that shows the metric. Hover over the panel, then click the bell icon at the top of the panel. 
```

ngrayluna commented on content/guides/core/automations/automation-events.md:
```diff
@@ -55,48 +55,56 @@ This section describes the events related to an artifact that can trigger an aut
+
+To set up a run metric automation, you configure how to compare the metric's value with the threshold you specify. Your choices depend on the event type and on any filters you specify. You can configure the threshold in the following ways:
+
+- **Run metrics threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged values meets the threshold you specify.
```
Unsure if we need to have "run" for this and "Run metrics change threshold met" (since it's nestled under "Run events"). 

The main take away is that we track metrics in an experiment. If and/or when they change, we trigger an automation. I think that message gets lost when we add "run" since it's a W&B-specific computation step and nothing to do with the actual ML stuff. 

```suggestion
- **Metrics threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged values meets the threshold you specify.
```

mdlinville replied:
![Screenshot 2025-06-06 at 2 18 38 PM](https://github.com/user-attachments/assets/b6733874-b9fc-4098-9443-f1e6feb4a952)
I think I need to call it the same as in the UI. WDYT?

ngrayluna commented on content/guides/core/automations/create-automations/slack.md:
```diff
@@ -61,12 +61,10 @@ A Registry admin can create automations in that registry.
-
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+1. Log in to W&B and go to the project page.
+1. In the sidebar, click **Automations**, then click **Create automation**. For a run metric, you can instead create an automation from a line plot for the metric. Hover over the panel, then click the bell icon at the top of the panel.
```
Similar comment here, "Run metric" seems to imply a metric about that run object itself, not the stuff we are logging to it.

```suggestion
1. In the sidebar, click **Automations**, then click **Create automation**. For a metric, you can instead create an automation from a line plot that shows the metric. Hover over the panel, then click the bell icon at the top of the panel.
```

mdlinville replied:
I can't spot the change your suggestion is making.

mdlinville commented on content/guides/core/automations/_index.md:
```diff
@@ -17,15 +17,16 @@ For example, an automation can notify a Slack channel when a new version is crea
-- When a metric for a run changes by a configured threshold.
+- When a metric for a run meets or exceeds a configured threshold. Optionally filter runs by user or run name.
+- When a metric for a run changes by a configured threshold. Optionally filter runs by user or run name.
+- When a run's status changes to **Running**, **Failed**, or **Finished**. Optionally filter runs by user or run name.
```
```suggestion
- When a run's status changes to **Running**, **Failed** (including jobs that were interactively killed), or **Finished**. Optionally filter runs by user or run name.
```

mdlinville committed:
Copyedits after rebase

ngrayluna commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -55,21 +55,32 @@ This section describes the events related to an artifact that can trigger an aut
+{{% /alert %}}
 
-To set up a run metric automation, you configure how to compare the metric's value with the threshold you specify. Your choices depend on the event type and on any filters you specify.
+Trigger a workflow when a run changes its status to one or more of **Running**, **Finished**, or **Failed**. **Finished** indicates a run that completed successfully. Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
```
This makes it sound like a run can have more than one status. Maybe


```suggestion
Trigger a workflow when a run changes its [status]({{< relref "/guides/runs/_index.md#run-states" >}}) to **Running**, **Finished**, or **Failed**. Optionally, you can further limit the runs that can trigger an automation by filtering by the user that started a run or the run's name.
```

Note: Update link to relref

ngrayluna commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -83,7 +94,7 @@ For **Run metrics threshold met** events, you configure:
 For example, trigger an automation when average `accuracy` is above `.6`.
 
-#### Change threshold
+##### Change threshold
```
This font looks really small...maybe we bump up the size in the CSS?

mdlinville replied:
We rarely need an H5. I dithered just making them bold but thought it would be useful to have a section-level link. Let me see how the sizes are defined and see if it is feasible to adjust it.

mdlinville replied:
Yeah that's what I was afraid of, we really don't have much room to adjust the size of H5. I suspect that H4-H5 is more important to get right than H5-H6, visually. WDYT?

```css
// Heading sizes
h1, .td-content h1 { font-size: 32px !important; }
h2, .td-content h2 { font-size: 24px !important; }
h3, .td-content h3 { font-size: 20px !important; }
h4, .td-content h4 { font-size: 16px !important; }
h5, .td-content h5 { font-size: 14px !important; }
h6, .td-content h6 { font-size: 12px !important; }
```

mdlinville replied:
https://github.com/wandb/docs/pull/1329/files#diff-1eea53605e04631cf5b5307d1f27eb125121316a58c492532d937130b2836e8fR145

ngrayluna commented on content/en/guides/core/automations/create-automations/slack.md:
```diff
@@ -61,12 +61,10 @@ A Registry admin can create automations in that registry.
-
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
+1. Log in to W&B and go to the project page.
+1. In the sidebar, click **Automations**, then click **Create automation**. For a metric, you can instead create an automation from a line plot that shows the metric. Hover over the panel, then click the bell icon at the top of the panel.
```
"For a metric..." confused me. Let's use the same  (albeit verbatim) terms.

```suggestion
1. In the sidebar, click **Automations**, then click **Create automation**. For a [run metrics change]({{< relref "/guides/core/automations/automation-events.md#run-metrics-change" >}}), you can instead create an automation from a line plot that shows the metric. Hover over the panel, then click the bell icon at the top of the panel.
```

Link to run metric change: https://docs-1453.docodile.pages.dev/guides/core/automations/automation-events/#run-metrics-change

ngrayluna commented on content/en/guides/core/automations/create-automations/webhook.md:
```diff
@@ -64,10 +64,8 @@ A Registry admin can create automations in that registry. Registry automations a
-
-    Or, from a line plot in the workspace, you can quickly create a [run metric automation]({{< relref "/guides/core/automations/automation-events.md#run-events" >}}) for the metric it shows. Hover over the panel, then click the bell icon at the top of the panel.
-    {{< img src="/images/automations/run_metric_automation_from_panel.png" alt="Illustration showing the location of the automation bell icon" >}}
+1. In the sidebar, click **Automations**, then click **Create automation**. For a metric, you can instead create an automation from a line plot that shows the metric. Hover over the panel, then click the bell icon at the top of the panel.
```
Same comment applies here. "For a metric..." is confusing...

mdlinville replied:
```suggestion
1. In the sidebar, click **Automations**, then click **Create automation**. For a [run metrics change]({{< relref "/guides/core/automations/automation-events.md#run-metrics-change" >}}), you can instead create an automation from a line plot that shows the metric. Hover over the panel, then click the bell icon at the top of the panel.
```

ngrayluna reviewed (commented):
Some minor comments. 

Also, the link at the bottom of this [_index.md](https://docs-1453.docodile.pages.dev/guides/core/automations/) page (In Limitations) redirects to [/guides/core/automations/automation-events/#run-metrics-events](https://docs-1453.docodile.pages.dev/guides/core/automations/automation-events/#run-metrics-events) but that doesn't exit.  I think we mean to redirect to  [/guides/core/automations/automation-events/#run-metrics-change](https://docs-1453.docodile.pages.dev/guides/core/automations/automation-events/#run-metrics-change)?

mdlinville committed:
Merge remote-tracking branch 'origin/main' into DOCS-1453

mdlinville committed:
Apply suggestions from code review

Co-authored-by: Noah Luna <15202580+ngrayluna@users.noreply.github.com>

mdlinville committed:
Update content/en/guides/core/automations/create-automations/webhook.md

mdlinville committed:
Merge remote-tracking branch 'origin/DOCS-1453' into DOCS-1453

mdlinville committed:
Fix link

mdlinville committed:
Adjust font size of H4-H6

Currently, H5 and H6 are each smaller than the content base font of 16px
and this looks off.

mdlinville commented on assets/scss/_variables_project.scss:
```diff
@@ -142,9 +142,9 @@ h1, h2, h3, h4, h5, h6 {
-h4, .td-content h4 { font-size: 16px !important; }
-h5, .td-content h5 { font-size: 14px !important; }
-h6, .td-content h6 { font-size: 12px !important; }
+h4, .td-content h4 { font-size: 18px !important; }
```
@johndmulhausen PTAL, context at https://github.com/wandb/docs/pull/1329/files#r2162679831.

johndmulhausen replied:
I think it looks give how you have it here :) The preview looks like you're striking the best balance you can. I leave Noah to do the final docs team approval though.

mdlinville committed:
Merge branch 'main' into DOCS-1453

trane293 commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -55,21 +55,32 @@ This section describes the events related to an artifact that can trigger an aut
 - **Adding a new alias to a version of an artifact**: Trigger a specific step of your workflow when a new artifact version in a project or collection has a specific label or alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.
 
 ### Run events
```
We should add a UI screenshot here

mdlinville replied:
I decided not to add a screenshot for this for now.

trane293 commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -55,21 +55,32 @@ This section describes the events related to an artifact that can trigger an aut
-The notification can trigger a workflow when these events occur:
-- **Run metrics threshold met**: When the average, minimum, or maximum logged value for a given metric meets the threshold you specify.
-- **Run metrics change threshold met**: When the average, minimum, or maximum logged value for a given metric changes by the threshold you specify.
+#### Run status change
```
We should add a UI screenshot here

mdlinville replied:
Added below

trane293 commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -83,7 +94,7 @@ For **Run metrics threshold met** events, you configure:
 For example, trigger an automation when average `accuracy` is above `.6`.
 
-#### Change threshold
+##### Change threshold
```
We should add a UI screenshot here

mdlinville replied:
Done

ngrayluna reviewed (approved):
Approving for tech content. Will defer to @anmolmann  for technical accuracy.

mdlinville committed:
Merge remote-tracking branch 'origin/main' into DOCS-1453

mdlinville committed:
Add images to Event Scopes page

mdlinville committed:
Add images to Slack and Webhook pages

mdlinville committed:
Align image dimensions

mdlinville commented on content/en/guides/core/automations/automation-events.md:
```diff
@@ -99,13 +118,15 @@ To create the automation, you configure:
+
+For example, trigger an automation when average `loss` decreases by at least `.25`.
+
+![Screenshot showing a run metrics change threshold automation](/images/automations/run_metrics_change_threshold_automation.png)
```
Also added a screenshot here

mdlinville commented on content/en/guides/core/automations/create-automations/slack.md:
```diff
@@ -22,8 +22,8 @@ A team admin can add a Slack integration to the team.
 
-    If necessary, sign in to Slack in your browser. When prompted, grant W&B permission to post to the Slack channel you select. Read the page, then click **Search for a channel** and begin typing the channel name. Select the channel from the list, then click **Allow**.
-
+    ![Screenshot showing two Slack integrations in a Team](/images/automations/slack_integrations.png)
```
Added a screenshot here

mdlinville commented on content/en/guides/core/automations/create-automations/webhook.md:
```diff
@@ -38,6 +38,8 @@ If the webhook requires a Bearer token or its payload requires a sensitive strin
 
     If the test does not succeed, verify the webhook's configuration and try again. If necessary, refer to [Troubleshoot your webhook]({{< relref "#troubleshoot-your-webhook" >}}).
 
+![Screenshot showing two webhooks in a Team](/images/automations/webhooks.png)
```
Added a screenshot here

github-actions[bot] committed:
Optimised images with calibre/image-actions

mdlinville committed:
Crop some  cruft from an image

mdlinville committed:
Fix image filename

mdlinville committed:
Merge remote-tracking branch 'origin/DOCS-1453' into DOCS-1453

github-actions[bot] committed:
Optimised images with calibre/image-actions

mdlinville committed:
Merge branch 'main' into DOCS-1453
